slsim/Halos/halos_lens.py

```diff
@@ -0,0 +1,1896 @@
+        :type samples_number: int, optional
+        :param mass_sheet: Flag to decide whether to use the mass_sheet correction. If set to False, the mass_correction_list is ignored. Defaults to True.
+        :type mass_sheet: bool, optional
+        """
+
+        if mass_correction_list is None:
+            mass_correction_list = {}
+            if mass_sheet:
+                warnings.warn(
+                    "Mass sheet correction is not applied but mass_sheet is set to True."
+                )
+                mass_sheet = False
+        if not mass_sheet:
+            mass_correction_list = {}
+            self.n_correction = 0
+            self.mass_sheet_correction_redshift = mass_correction_list.get("z", [])
+            #            self.mass_first_moment = mass_correction_list.get("first_moment", [])
+            self.mass_sheet_kappa = mass_correction_list.get("kappa", [])
+        if mass_sheet:
+            self.n_correction = len(mass_correction_list)
+            self.mass_sheet_correction_redshift = mass_correction_list["z"]
+            self.mass_sheet_kappa = mass_correction_list["kappa"]
+        self.z_source = z_source
+        self.halos_list = halos_list
+        self.mass_correction_list = mass_correction_list
+        self.mass_sheet = mass_sheet
+        if np.isnan(halos_list["z"][0]) and len(halos_list) == 1:
+            self.n_halos = 0
+        else:
+            self.n_halos = len(self.halos_list)
+        self.sky_area = sky_area
+        # if self.n_halos == 0:
+        #     self.halos_redshift_list = []
+        # else: (not righting this because of unpleaseant error dealing with redshift nan)
+        #    self.halos_redshift_list = halos_list["z"]
+        self.halos_redshift_list = halos_list["z"]
+        self.mass_list = halos_list["mass"]
+        self.samples_number = samples_number
+        self._z_source_convention = 5
+        # todo: z_source_convention set same with halos.py and yml
+        if cosmo is None:
+            warnings.warn(
+                "No cosmology provided, instead uses astropy.cosmology import default_cosmology"
+            )
+            import astropy.cosmology as cosmology
+
+            self.cosmo = cosmology.default_cosmology.get()
+        else:
+            self.cosmo = cosmo
+
+        self.combined_redshift_list = np.concatenate(
+            (self.halos_redshift_list, self.mass_sheet_correction_redshift)
+        )
+
+        self._lens_cosmo = None  # place-holder for lazy load
+        self._lens_model = None  # same as above
+        c_200 = [
+            concentration_from_mass(z=zi, mass=mi)
+            for zi, mi in zip(self.halos_redshift_list, self.mass_list)
+        ]
+        self.halos_list["c_200"] = c_200
+        self.enhance_halos_table_random_pos()
+
+        # TODO: Set z_source as an input parameter or other way
+
+    @property
+    def lens_cosmo(self):
```

Comment:
this might be confusing, as this is a list of lens_cosmo class instances. Perhaps re-name it and document it properly